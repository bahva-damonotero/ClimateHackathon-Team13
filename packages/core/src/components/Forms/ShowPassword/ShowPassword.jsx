import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import showPassword from '@uswds/uswds/js/_usa-password';

export default function ShowPassword({ target }) {
  const containerRef = useRef();

  useEffect(() => {
    const containerEl = containerRef.current;
    if (!containerEl) return;

    // call USWDS's `on` function to initialize the component
    showPassword.on(containerEl);

    // on dismount, call USWDS's `off` function to cleanup the component
    return () => {
      showPassword.off(containerEl);
    };
  }, []);

  const multiple = Array.isArray(target);
  const ariaControls = multiple ? target.join(' ') : target;

  return (
    <span className="bah-uswds-show-password" ref={containerRef}>
      <button
        type="button"
        aria-controls={ariaControls}
        className="usa-show-password"
        data-show-text={multiple ? 'Show my typing' : 'Show password'}
        data-hide-text={multiple ? 'Hide my typing' : 'Hide password'}
      >
        {multiple ? 'Show my typing' : 'Show password'}
      </button>
    </span>
  );
}

ShowPassword.propTypes = {
  /** The ID or ID(s) of the input(s) to target */
  target: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ])
};
