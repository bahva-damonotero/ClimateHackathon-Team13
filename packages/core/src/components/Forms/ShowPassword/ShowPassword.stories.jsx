import React from 'react';

import ShowPassword from '.';
import { Form, FormNote, Input } from '..';

export default {
  title: 'Forms/ShowPassword',
  component: ShowPassword,
  subcomponents: {
    Input
  }
};

const OneFieldTemplate = (args) => (
  <Form>
    <Input id="password" type="password">
      Password
    </Input>

    <FormNote>
      <ShowPassword {...args} />
    </FormNote>
  </Form>
);

export const oneField = OneFieldTemplate.bind({});
oneField.args = {
  target: 'password'
};

const TwoFieldsTemplate = (args) => (
  <Form>
    <Input id="newPass" type="password">
      New Password
    </Input>

    <Input id="confirmPass" type="password">
      Confirm Password
    </Input>

    <FormNote>
      <ShowPassword {...args} />
    </FormNote>
  </Form>
);

export const twoFields = TwoFieldsTemplate.bind({});
twoFields.args = {
  target: ['newPass', 'confirmPass']
};
