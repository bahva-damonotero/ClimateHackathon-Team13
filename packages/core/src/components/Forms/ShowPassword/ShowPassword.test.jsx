import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import showPassword from '@uswds/uswds/js/_usa-password';

import ShowPassword from '.';
import { Form, FormNote, Input } from '..';

describe('ShowPassword', () => {
  const singleFieldJsx = (
    <Form>
      <Input id="password" type="password">
        Password
      </Input>

      <FormNote>
        <ShowPassword target="password" />
      </FormNote>
    </Form>
  );

  it('should call `showPassword.on()` to initialize the component', () => {
    const onSpy = jest.spyOn(showPassword, 'on');

    render(singleFieldJsx);

    expect(onSpy).toHaveBeenCalled();
  });

  it('should match snapshot (single field)', () => {
    const tree = renderer.create(singleFieldJsx).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot (two fields)', () => {
    const tree = renderer
      .create(
        <Form>
          <Input id="newPass" type="password">
            New Password
          </Input>

          <Input id="confirmPass" type="password">
            Confirm Password
          </Input>

          <FormNote>
            <ShowPassword target={['newPass', 'confirmPass']} />
          </FormNote>
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
