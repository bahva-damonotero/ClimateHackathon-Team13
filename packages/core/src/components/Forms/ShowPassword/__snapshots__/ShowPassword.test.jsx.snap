// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowPassword should match snapshot (single field) 1`] = `
<form
  className="usa-form"
>
  <label
    className="usa-label"
    htmlFor="password"
  >
    Password
  </label>
  <input
    className="usa-input"
    id="password"
    name="password"
    type="password"
  />
  <p
    className="usa-form__note"
  >
    <span
      className="bah-uswds-show-password"
    >
      <button
        aria-controls="password"
        className="usa-show-password"
        data-hide-text="Hide password"
        data-show-text="Show password"
        type="button"
      >
        Show password
      </button>
    </span>
  </p>
</form>
`;

exports[`ShowPassword should match snapshot (two fields) 1`] = `
<form
  className="usa-form"
>
  <label
    className="usa-label"
    htmlFor="newPass"
  >
    New Password
  </label>
  <input
    className="usa-input"
    id="newPass"
    name="newPass"
    type="password"
  />
  <label
    className="usa-label"
    htmlFor="confirmPass"
  >
    Confirm Password
  </label>
  <input
    className="usa-input"
    id="confirmPass"
    name="confirmPass"
    type="password"
  />
  <p
    className="usa-form__note"
  >
    <span
      className="bah-uswds-show-password"
    >
      <button
        aria-controls="newPass confirmPass"
        className="usa-show-password"
        data-hide-text="Hide my typing"
        data-show-text="Show my typing"
        type="button"
      >
        Show my typing
      </button>
    </span>
  </p>
</form>
`;
